core/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java

```diff
@@ -0,0 +1,735 @@
+    }
+
+    @Override
+    public RangeFieldType fieldType() {
+        return (RangeFieldType) super.fieldType();
+    }
+
+    @Override
+    protected String contentType() {
+        return fieldType.typeName();
+    }
+
+    @Override
+    protected RangeFieldMapper clone() {
+        return (RangeFieldMapper) super.clone();
+    }
+
+    @Override
+    protected void parseCreateField(ParseContext context, List<IndexableField> fields) throws IOException {
+        final boolean includeInAll = context.includeInAll(this.includeInAll, this);
+        Range range;
+        if (context.externalValueSet()) {
+            range = context.parseExternalValue(Range.class);
+        } else {
+            XContentParser parser = context.parser();
+            if (parser.currentToken() == XContentParser.Token.START_OBJECT) {
+                RangeFieldType fieldType = fieldType();
+                String fieldName = null;
+                Number from = null;
+                Number fromNullValue = fieldType.rangeType.minValue();
+                Number to = null;
+                Number toNullValue = fieldType.rangeType.maxValue();
+                boolean includeFrom = DEFAULT_INCLUDE_LOWER;
+                boolean includeTo = DEFAULT_INCLUDE_UPPER;
+                LongSupplier nowInMillis = () -> System.currentTimeMillis();
+                XContentParser.Token token;
+                while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+                    if (token == XContentParser.Token.FIELD_NAME) {
+                        fieldName = parser.currentName();
+                    } else {
+                        if (fieldName.equals(GT_FIELD.getPreferredName())) {
+                            from = fieldType.parseNumericValue(parser, fromNullValue, coerce.value(), nowInMillis);
+                            includeFrom = false;
+                        } else if (fieldName.equals(GTE_FIELD.getPreferredName())) {
+                            from = fieldType.parseNumericValue(parser, fromNullValue, coerce.value(), nowInMillis);
+                            includeFrom = true;
+                        } else if (fieldName.equals(LT_FIELD.getPreferredName())) {
+                            to = fieldType.parseNumericValue(parser, toNullValue, coerce.value(), nowInMillis);
+                            includeTo = false;
+                        } else if (fieldName.equals(LTE_FIELD.getPreferredName())) {
+                            to = fieldType.parseNumericValue(parser, toNullValue, coerce.value(), nowInMillis);
+                            includeTo = true;
+                        } else {
+                            throw new MapperParsingException("error parsing field [" +
+                                name() + "], with unknown parameter [" + fieldName + "]");
+                        }
+                    }
+                } // while
+                range = new Range(from, to, includeFrom, includeTo);
```

Comment:
I think we should either raise an exception if `from` or `to` is null, meaning that one of the bounds was not specified, or make their values default to the min/max values of the range type. I think the latter is better as it is more consistent with the range query?